### ISSUE TITLE
A bug in test compiler test class SubIfFalseThenMethod

### ISSUE THREAD

**Issue by GesusK** [eclipse/omr#1317]
One test class SubIfFalseThenMethod looks not implemented correctly. And the reason I post it here as a github issue is that we need to determine it's the user's job (in this case test writer) or IL generator's job to avoid such a problem.

When the function `bool SubIfFalseThenMethod::buildIL()` gets called, it calls
```c++ 
IfThenElse(NULL, &elsePath, ConstInt64(0));
```
inside which the section in `Then`, pointed by `NULL` (`&thenPath`), would never be reached or executed because of the if condition is a constant `0`. Therefore, when the IL generated for this part is represented by a block in the control flow graph, it has no entry or exit. This dangling basic block fails the CFG check because it has the exit block as a successor, but no return inside it. Below is the log for the compilation, and block_5 is the problematic block:
```
BBStart <block_2>                                                                   
treetop                                                                             
  lconst 0                                                                          
treetop                                                                             
  lconst 0                                                                          
iflcmpne --> block_5 BBStart at n9n ()                                              
  ==>lconst 0
  ==>lconst 0
BBEnd </block_2> =====                                                              

BBStart <block_6>                                                                   
BBEnd </block_6> =====                                                              

BBStart <block_3>                                                                   
treetop                                                                             
  lload  valueA<parm 0 J>[#612  Parm] [flags 0x40010104 0x0 ]                       
treetop                                                                             
  lload  valueB<parm 1 J>[#613  Parm] [flags 0x40010104 0x0 ]                       
treetop                                                                             
  lsub                                                                              
    ==>lload
    ==>lload
lreturn                                                                             
  ==>lsub
BBEnd </block_3> =====                                                              

BBStart <block_5>                                                                   
BBEnd </block_5>  ```

As stated at the beginning, a decision needs to be made on either the IL needs to address such an unreachable "then" section, or the user is responsible to avoid such an invalid input for the IL builder.

**Comment by mstoodle** [comment]
Interesting discovery; that looks like an IlBuilder bug to me. Clients of IlBuilder shouldn't need to worry about whether a service will create proper IL or not.

**Comment by Leonardo2718** [comment]
I completely agree that this is an IlBuilder bug for the same reason that @mstoodle stated.

As an additional note, generating improper IL for testing purposes should be done using something other than IlBuilder (e.g. Tril :slightly_smiling_face:).

Ideally, I would like the test compiler to not use IlBuilders at all... but I digress from the main issue.

**Comment by GesusK** [comment]
I'm on the same side, like it or not, a senseless if statement can be normal in real code.

So for this issue, my next step would be investigating the IL generation process to find how such a case should be represented.

**Comment by mgaudet** [comment]
So, on further reflection, and discussion with @GesusK, I am starting to think the problem is actually a test case error: 

```
bool
SubIfFalseThenMethod::buildIL()
   {
   TR::IlBuilder *elsePath = NULL;
   IfThenElse(NULL, &elsePath, ConstInt64(0));

   elsePath->Return(
   elsePath->   Sub(
   elsePath->      Load("valueA"),
   elsePath->      Load("valueB")));

   return true;
   }
```

If we rewrite this in a C-like syntax, I think the error stands out more clearly: 

```
Int64 SubIfFalseThenMethod(Int64 valueA, Int64 valueB) { 
   if (0) { 
   } else { 
     return valueA - valueB;
   } 
  // Merge point
}
```

The problem here is actually that the test case doesn't provide a valid return for the merge point, which is outside the responsibility of `IfThenElse`. 

Now, this is slightly awkward, because providing CFG correctness through the Builder pattern of JitBuilder in face of reasoning about returns is... maybe harder?

**Comment by mstoodle** [comment]
Hm, good point, @mgaudet . Thanks for digging deeper!

It's another case where the method can 'fall off' the end of the builder because an explicit Return was not done on some path (even an un-executable path).

I revise my statement: this is "invalid" builder input, so the test case should be updated to add a Return(-1) or something like that after the IfThenElse() . But I will certainly agree that the way the problem is communicated to the user is, well, unfortunate and not particularly helpful.

Possibly the thing that added a flow edge from block_5 to the `cfg()->getEnd()` could detect that the block did not have a return inserted into it, but my guess is that `block_5` is actually a pre-created "exit" block for the method builder and would never have any code and it just defaults to have an edge to `cfg()->getEnd()`.

Maybe we could construct an assertion for the last generated block ending in some real form of control flow (after unreachable blocks are removed)? That would provide a slightly better hint to the actual problem...

**Comment by GesusK** [comment]
I added three tests, which basically verified that a use case of `IfThenElseMethodBuilder` will pass CFG check only when either
1) both `thenPath` and `elsePath` have their `Return()` defined; or
2) at the end of IlBuilder, a `Return()` is defined after the branching.
This would require the JIT Builder user to always make sure there is a return at the end of the built method.
And as discussed with @mgaudet , a point came up that we might need to always automatically generate a default return value in MethodBuilder. For example, if the method being built requires a return type of `int`, MethodBuilder will then automatically insert a default value with type `int` at the end of the method.

**Comment by mgaudet** [comment]
By ensuring every method builder already has a return, we can verify that the CFG requirement base case is preserved. It may not be a sensible return, but would make a method builder into a sort of template like this: 

```
ReturnType MethodBuilder(...) { 
   // USER CODE GOES HERE
   return DefaultValue<ReturnType>();
}
```

It may be worth while to also generate an abort in there, to make sure the user defined code actually returns on all paths.

**Comment by GesusK** [comment]
Turn on Tree verifier: https://github.com/GesusK/omr/commit/6621b1f7d982efb2d4c09299c93b458c9f81bc0c

Add new tests: https://github.com/GesusK/omr/commit/c9b1393b3e9cf12ab632067dc5fdef5b6bbc015a

**Comment by mstoodle** [comment]
I would actually prefer an assert to a default return. I am happier that the test case crashed, even in a confusing way, rather than producing code, because it reports a problem to the user rather than silently doing something that is unlikely to be what is actually intended. If we make this default return value approach the way it works, then some people actually will create MethodBuilders that do it intentionally as well as making it harder for the unintentional victims to discover their mistake.

The scenario where control falls off the end of the method is something we should be able to detect. With only a few seconds thought (meaning I may have missed some subtle scenarios), I think asserting at the end of buildIL() (i.e. when it comes back in MethodBuilder::injectIL() ) that the method builder object has setDoesNotComeBack() should do the trick. I think we'll have to propagate that flag setting when AppendBuilder() is called.

Alternatively, at the same point in injectIL(), we could walk backwards through _sequence to verify that the last real opcode is either a goto, a lookup, or a return, and that there are no other incoming edges.

**Comment by GesusK** [comment]
@mstoodle That means the JITBuilder will just take the responsibility of verifying whether or not there is a return value, while the user will always be responsible for providing return values, right?
I see the point of having an automatically generated return value being improved robustness, and less requirements on user. But simply adding an assertion seems making the method building process more accurately and explicitly.
Feeling confused and hesitating.

**Comment by mstoodle** [comment]
"I see the point of having an automatically generated return value being improved robustness, and less requirements on user"

I disagree. Automatically generating a return statement is filling in a blank the user left without any way to know what the user really meant (so return some arbitrary default value). But remember the user maybe just forgot to insert "Return(theRealResult)". So returning a default value in that scenario could be *really* incorrect. Imagine, for example, the function is called "bankBalanceAfterDepositingMyCheque()" and we return the default value zero. Super not good!

In my opinion, JitBuilder's responsibility here is to tell the user something wasn't done properly so the user can fix it appropriately (i.e. return the correct result).

**Comment by Leonardo2718** [comment]
I agree with @mstoodle. Returning a default value feels like saying "program is ill-formed; no diagnostic required". This is a known source of frustration in C++ (see [here](http://en.cppreference.com/w/cpp/language/return#Notes) and [here](https://stackoverflow.com/questions/20614282/why-does-this-c-snippet-compile-non-void-function-does-not-return-a-value)). Using an assert instead clearly tells the user that something's wrong.
